### PR TITLE
Add context benchmarks

### PR DESCRIPTION
**Pull request by dfawley**

### CODE REVIEW

**Comment by dfawley** [comment]
```
$ go test -benchtime=3s -bench=Context benchmark/primitives/*.go
goos: linux
goarch: amd64
BenchmarkCancelContextErrNoErr-12            	100000000	        52.2 ns/op
BenchmarkCancelContextErrGotErr-12           	100000000	        51.9 ns/op
BenchmarkCancelContextChannelNoErr-12        	200000000	        19.9 ns/op
BenchmarkCancelContextChannelGotErr-12       	100000000	        36.1 ns/op
BenchmarkTimerContextErrNoErr-12             	100000000	        53.1 ns/op
BenchmarkTimerContextErrGotErr-12            	100000000	        52.6 ns/op
BenchmarkTimerContextChannelNoErr-12         	200000000	        20.5 ns/op
BenchmarkTimerContextChannelGotErr-12        	100000000	        37.0 ns/op
BenchmarkCancelNetContextErrNoErr-12         	100000000	        51.8 ns/op
BenchmarkCancelNetContextErrGotErr-12        	100000000	        51.7 ns/op
BenchmarkCancelNetContextChannelNoErr-12     	200000000	        20.3 ns/op
BenchmarkCancelNetContextChannelGotErr-12    	100000000	        36.1 ns/op
BenchmarkTimerNetContextErrNoErr-12          	100000000	        52.5 ns/op
BenchmarkTimerNetContextErrGotErr-12         	100000000	        53.4 ns/op
BenchmarkTimerNetContextChannelNoErr-12      	200000000	        20.5 ns/op
BenchmarkTimerNetContextChannelGotErr-12     	100000000	        36.7 ns/op
```

**Comment by dfawley** [comment]
I just realized there's no point in benchmarking `x/net/context`, at least not with more work:

- For Go versions > 1.6, `x/net/context` is implemented by `context` (and as of 1.9 it is a type alias)
- For Go 1.6, this benchmark cannot be run, because it imports `context`

File: benchmark/primitives/context_test.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	for i := 0; i < b.N; i++ {
+		if err := ctx.Err(); err != nil {
+			b.Fatal("error")
+		}
+	}
+	cancel()
+}
+
+func BenchmarkTimerContextErrGotErr(b *testing.B) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Microsecond)
```
Comment (MakMukhi) [inline benchmark/primitives/context_test.go]: Perhaps nanosecond? Or maybe sleep for a microsecond to be sure?

Comment (dfawley) [reply benchmark/primitives/context_test.go]: I still call cancel on the next line..  I don't expect that deadline exceeded vs. cancellation results in different paths through `.Err()` or `.Done()`, but `TimerContext` is a different implementation from `CancelContext` which is why I have two benchmarks.

Comment (MakMukhi) [reply benchmark/primitives/context_test.go]: Yeah you're right; end the of the day it's only a cancelled context that we're checking for.

**Pull request merged**